jyson supports an array of objects

earobinson opened the pull request:
- Added in testing for node 8

Resolves Issue #9

earobinson committed:
jyson supports an array of objects

- Added in testing for node 8

Resolves Issue #9

earobinson committed:
Refactored the way that we handle arrays

- We now ues a while loop till we get to the end of the elements

earobinson commented on lib/jyson.js:
```diff
@@ -1,19 +1,13 @@
-  // We have found our key
-  if (arraySplitKeys.length === 0) {
-    return value;
+    assert(_.get(object, arrayKey).length > opts.arrayIndex['$'], 'We hit the end of the array');
```
Not sure how I feel about this

etorrejon replied:
Looks like this function returns a value and not a key? If that's the case you may want to rename it to `getValue` or something similar.

etorrejon replied:
How is `opts.arrayIndex` used?  Could the `arrayIndex` value be stored outside of `opts`? It seems that `opts` stands for "options", but it's being used to track current index?

earobinson replied:
> Looks like this function returns a value and not a key? If that's the case you may want to rename it to `getValue` or something similar.

If you are referring to `getKey` it returns a value

earobinson replied:
> How is `opts.arrayIndex` used?  Could the `arrayIndex` value be stored outside of `opts`? It seems that `opts` stands for "options", but it's being used to track current index?

@etorrejon great point. `opts` is passed to functions https://github.com/hubba/jyson/blob/master/spec/lib/jyson/jyson.function.spec.js#L52 as they will need to to get the correct data.

Do you think we should rename opts? Should we add an additional argument to functions?

earobinson replied:
@misshula @etorrejon My original comment however was about the `assert` Im not sure that relying on an error to exit out of a loop is a good pattern. Thoughts?

etorrejon replied:
@earobinson 
- yup, I was referring to `getKey`. `getValue` would be more intuitive.
- my question about the `arrayKey` was whether it actually needs to be part of `opts`. It seems as though it doesn't, since it's used to track an array index and not a configuration option?

etorrejon replied:
also... re: your original question, I think it's fine to error out in that case

etorrejon replied:
... if jyson were attempting to read an index that was out of bounds, we'd likely want to know.

earobinson replied:
> ... if jyson were attempting to read an index that was out of bounds, we'd likely want to know.

@etorrejon 

Issue is we catch the error here https://github.com/hubba/jyson/pull/11/files#diff-939716d327873b484c0914e8a74fc40bR52

misshula commented on spec/lib/jyson/jyson.array.spec.js:
```diff
@@ -15,41 +15,216 @@ describe('jyson.array.spec: an array in the template', () => {
+              {
+                b: {
+                  c: 'easy as 123',
+                  d: 'do ray me'
```
do re mi

etorrejon commented on changelog.md:
```diff
@@ -0,0 +1,4 @@
+# jyson changelog
+
+## v1.2.0
+- jyson now supports and array of objects Issue [#9](//hubba/jyson/issues/9)
```
minor: typo `and` should be `an`

etorrejon commented on lib/jyson.js:
```diff
@@ -72,7 +86,7 @@ module.exports.buildTemplateFunction = (templateObject, buildTemplateFunctionOpt
       const json = _.cloneDeep(templateObject);
 
-      return fillKeys(json, object, buildTemplateFunctionOpts.undefinedValue, opts);
+      return fillKeys(json, object, buildTemplateFunctionOpts.undefinedValue, Object.assign(opts, {arrayIndex: {'$': 0}}));
```
What's the difference between `json` and `object`?

earobinson replied:
`json` is the json we are generating
`object` is the object we are generating the json from

Open to better names / documentation.

etorrejon commented on spec/lib/jyson/jyson.array.spec.js:
```diff
@@ -15,41 +15,216 @@ describe('jyson.array.spec: an array in the template', () => {
+          expect(json.a[1].c).to.equal('a1c');
+        });
+
+        it('must handel undefined values an object to json', () => {
```
Suggest changing spec to `must convert undefined values to null`? Not sure if that's the intention here

etorrejon reviewed (dismissed):
A few comments

earobinson committed:
Addressing comments:

- https://github.com/hubba/jyson/pull/11#discussion_r140855354
- https://github.com/hubba/jyson/pull/11#discussion_r140864246

misshula reviewed (approved):
let's see how this goes and make the necessary changes later...
